File: benches/comparative/typescript/bench_kalshi.mjs
// Kalshi REST API head-to-head: OpenPX vs kalshi-typescript-sdk.
//
// 20 iterations × 100 ms gap. Writes JSON summary to stdout.
//
// Run: `node bench_kalshi.mjs > ../results/typescript_kalshi.json`

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { Bench } from "tinybench";
import { Exchange } from "@openpx/sdk";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const META = JSON.parse(
  fs.readFileSync(
    path.join(__dirname, "..", "fixtures", "kalshi_orderbook.meta.json"),
    "utf8"
  )
);
const TICKER = META.ticker;
const GAP_MS = 100;

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

// Try the official Kalshi SDK; fall back to OpenPX-only.
let sdkClient = null;
let sdkName = null;
for (const pkg of [
  "kalshi-typescript",
  "kalshi-typescript-sdk",
  "kalshi-ts",
  "@kalshi/sdk",
]) {
  try {
    const mod = await import(pkg);
    sdkClient = mod.default ?? mod.KalshiClient ?? mod.MarketsApi ?? mod;
    sdkName = pkg;
    break;
  } catch {
    /* try next */
  }
}
if (!sdkClient) {
  console.error("note: no kalshi SDK installed, OpenPX-only run");
}

const openpx = new Exchange("kalshi", {});
const bench = new Bench({ time: 0, iterations: 20 });

bench.add("openpx::kalshi::fetch_markets", async () => {
  await sleep(GAP_MS);
  await openpx.fetchMarkets();
});

bench.add("openpx::kalshi::fetch_market", async () => {
  await sleep(GAP_MS);
  await openpx.fetchMarket(TICKER);
});

bench.add("openpx::kalshi::fetch_orderbook", async () => {
  await sleep(GAP_MS);
  await openpx.fetchOrderbook(TICKER);
});

bench.add("openpx::kalshi::fetch_trades", async () => {
  await sleep(GAP_MS);
  await openpx.fetchTrades(TICKER);
});

// Kalshi SDK shapes vary across packages; resolve methods defensively.
if (sdkClient) {
  const tagged = (op) => `${sdkName}::kalshi::${op}`;

  bench.add(tagged("fetch_markets"), async () => {
    await sleep(GAP_MS);
    if (sdkClient.getMarkets) return sdkClient.getMarkets({ limit: 100 });
    if (sdkClient.markets?.getMarkets)
      return sdkClient.markets.getMarkets({ limit: 100 });
  });

  bench.add(tagged("fetch_market"), async () => {
    await sleep(GAP_MS);
    if (sdkClient.getMarket) return sdkClient.getMarket({ ticker: TICKER });
    if (sdkClient.markets?.getMarket)
      return sdkClient.markets.getMarket({ ticker: TICKER });
  });

  bench.add(tagged("fetch_orderbook"), async () => {
    await sleep(GAP_MS);
    if (sdkClient.getMarketOrderbook)
      return sdkClient.getMarketOrderbook({ ticker: TICKER });
    if (sdkClient.markets?.getMarketOrderbook)
      return sdkClient.markets.getMarketOrderbook({ ticker: TICKER });
  });

  bench.add(tagged("fetch_trades"), async () => {
    await sleep(GAP_MS);
    if (sdkClient.getTrades)
      return sdkClient.getTrades({ ticker: TICKER, limit: 100 });
    if (sdkClient.markets?.getTrades)
      return sdkClient.markets.getTrades({ ticker: TICKER, limit: 100 });
  });
}

await bench.run();

const tasks = bench.tasks.map((t) => ({
  name: t.name,
  mean_ns: t.result?.mean ? t.result.mean * 1e6 : null,
  stddev_ns: t.result?.sd ? t.result.sd * 1e6 : null,
  samples: t.result?.samples?.length ?? 0,
}));

console.log(JSON.stringify({ tasks }, null, 2));

File: benches/comparative/python/bench_kalshi.py
"""Kalshi REST API head-to-head: OpenPX vs kalshi-python.

20 iterations × 100 ms gap, polyfill methodology. Same series fixture
captured by `tools/capture_bench_fixtures.py`.

Run: `pytest benches/comparative/python/bench_kalshi.py --benchmark-only`
"""

from __future__ import annotations

import json
import time
from pathlib import Path

import pytest

from openpx import Exchange

try:
    from kalshi_python.api_instance_factory import ApiInstanceFactory
    from kalshi_python.configuration import Configuration

    HAS_KALSHI_PY = True
except ImportError:  # pragma: no cover - optional
    HAS_KALSHI_PY = False


FIXTURE = Path(__file__).resolve().parents[1] / "fixtures" / "kalshi_orderbook.meta.json"
GAP = 0.1


def _ticker() -> str:
    return json.loads(FIXTURE.read_text())["ticker"]


def _event_ticker() -> str:
    return json.loads(FIXTURE.read_text())["event_ticker"]


@pytest.fixture(scope="module")
def openpx_exchange() -> Exchange:
    return Exchange("kalshi")


@pytest.fixture(scope="module")
def kalshi_api():
    if not HAS_KALSHI_PY:
        pytest.skip("kalshi-python not installed")
    cfg = Configuration()
    # ApiInstanceFactory exposes per-resource clients (MarketApi, etc.)
    return ApiInstanceFactory(cfg)


# ---------------------------------------------------------------------------
# fetch_markets
# ---------------------------------------------------------------------------


def test_openpx_fetch_markets(benchmark, openpx_exchange):
    def run():
        time.sleep(GAP)
        return openpx_exchange.fetch_markets()

    benchmark.pedantic(run, iterations=1, rounds=20)


def test_kalshi_python_fetch_markets(benchmark, kalshi_api):
    market_api = kalshi_api.get_market_api()

    def run():
        time.sleep(GAP)
        return market_api.get_markets(limit=100)

    benchmark.pedantic(run, iterations=1, rounds=20)


# ---------------------------------------------------------------------------
# fetch_market (single)
# ---------------------------------------------------------------------------


def test_openpx_fetch_market(benchmark, openpx_exchange):
    t = _ticker()

    def run():
        time.sleep(GAP)
        return openpx_exchange.fetch_market(t)

    benchmark.pedantic(run, iterations=1, rounds=20)


def test_kalshi_python_fetch_market(benchmark, kalshi_api):
    market_api = kalshi_api.get_market_api()
    t = _ticker()

    def run():
        time.sleep(GAP)
        return market_api.get_market(ticker=t)

    benchmark.pedantic(run, iterations=1, rounds=20)


# ---------------------------------------------------------------------------
# fetch_orderbook
# ---------------------------------------------------------------------------


def test_openpx_fetch_orderbook(benchmark, openpx_exchange):
    t = _ticker()

    def run():
        time.sleep(GAP)
        return openpx_exchange.fetch_orderbook(t)

    benchmark.pedantic(run, iterations=1, rounds=20)


def test_kalshi_python_fetch_orderbook(benchmark, kalshi_api):
    market_api = kalshi_api.get_market_api()
    t = _ticker()

    def run():
        time.sleep(GAP)
        return market_api.get_market_orderbook(ticker=t)

    benchmark.pedantic(run, iterations=1, rounds=20)


# ---------------------------------------------------------------------------
# fetch_trades
# ---------------------------------------------------------------------------


def test_openpx_fetch_trades(benchmark, openpx_exchange):
    t = _ticker()

    def run():
        time.sleep(GAP)
        return openpx_exchange.fetch_trades(t)

    benchmark.pedantic(run, iterations=1, rounds=20)


def test_kalshi_python_fetch_trades(benchmark, kalshi_api):
    market_api = kalshi_api.get_market_api()
    t = _ticker()

    def run():
        time.sleep(GAP)
        return market_api.get_trades(ticker=t, limit=100)

    benchmark.pedantic(run, iterations=1, rounds=20)
